web3/hardhat/contracts/Crowdfund.sol

```diff
@@ -0,0 +1,134 @@
+import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+error NotEnoughETHDonation();
+error NotEnoughBalance();
+error FailedToSendNativeToken();
+
+contract Crowdfund is Ownable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    IERC20 public token;
+    uint256 public startsAt;
+    uint256 public endsAt;
+    uint256 public goal;
+    bytes public metaPtr;
+
+    bool public canRefund;
+
+    mapping(address => uint256) userFunds;
+
+    event donated(address sender, uint256 amount);
+    event funded(address to, uint256 amount);
+    event refundActive(bool refunding);
+    event userBalanceWithdrawn(address user, uint256 balance);
+
+    modifier crowfundingEnded() {
+        require(!hasCrowdfundingEnded(), "Crowd funding hasn't ended");
+        _;
+    }
+
+    constructor(bytes memory meta) {
+        // pass data to constructor for new matchingPool
+        (
+            address safe,
+            address _token,
+            uint256 _startsAt,
+            uint256 _endsAt,
+            uint256 _goal,
+            bytes memory _metaPtr
+        ) = abi.decode(
+                meta,
+                (address, address, uint256, uint256, uint256, bytes)
+            );
+
+        _transferOwnership(safe);
+        if (_token != address(0)) {
+            token = IERC20(_token);
+        }
+        startsAt = _startsAt;
+        endsAt = _endsAt;
+        goal = _goal;
+        metaPtr = _metaPtr;
+    }
+
+    // editCrowdfund
+
+    function donate(uint256 amount) public payable {
+        uint256 _donatedAmount = msg.value;
+        if (tokenIsNative()) {
+            // it's an ETH donation
+            if (msg.value == 0) {
+                revert NotEnoughETHDonation();
+            }
+        } else {
+            // it's ERC20 donation
+            // transfer amount to this contract
+            token.safeTransferFrom(msg.sender, address(this), amount);
+            _donatedAmount = amount;
+        }
+
+        // increase user balance
+        userFunds[msg.sender] += _donatedAmount;
+
+        emit donated(msg.sender, _donatedAmount);
+    }
+
+    // enableRefund
+    function enableRefund() public onlyOwner crowfundingEnded {
+        canRefund = true;
+
+        emit refundActive(true);
+    }
+
+    // fund
+    function fund() public onlyOwner nonReentrant {
+        uint256 amount = address(this).balance;
+        address to = owner();
+        if (tokenIsNative()) {
+            // withdraw native
+            (bool sent, ) = to.call{value: amount}("");
+            require(sent, "Failed to send Ether");
+        } else {
+            amount = token.balanceOf(address(this));
+            token.safeTransferFrom(address(this), to, amount);
+        }
+
+        emit funded(to, amount);
+    }
+
+    // withdraw
+    function withdraw() public nonReentrant {
```

Comment:
Do we want to enable withdraws like this? If so why?

Reply:
Resolved earlier